Added In-app list for failed library updates

Programmer-0-0 opened the pull request:
This should finally (and hopefully) close #4616. Here are some images about the newly added feature:

<details>
<summary>Details</summary>

| The warning icon is disabled and not visible by default but when an update job is run, and an error is detected it will become enabled| For now I only did it as a list (no grid view) and the migrate button has the same function as going to the migrate source screen|
| -------------   | ------------- |
![WhatsApp Image 2023-08-02 at 22 24 46](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/9c1d214e-a4a4-40fa-b438-405eb2ec7c9d)|                    ![WhatsApp Image 2023-08-02 at 22 24 46](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/df7b604c-26d0-47db-a455-d08194c2ed7a)
| I took @T3sT3ro advice and made the Pills to display the error count along side with the total count of mangas in that source| In the bottom bar there is a delete button as well as a dismiss button|
![WhatsApp Image 2023-08-02 at 22 24 48](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/d6ecd086-e9ad-4433-9abd-1c6c98f65e76) |        ![WhatsApp Image 2023-08-02 at 22 24 50](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/61f03a35-c6eb-4630-b0f9-00d1e7dc2a10)    
| There is also an option to group by category but I couldn't find a way to efficiently get the total manga count in that category   |I edited the EmptyScreen function to display happy faces when you dismiss or ignore all the errors|
![WhatsApp Image 2023-08-02 at 22 34 39](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/9f8588a4-037b-4fba-aebe-c2aaedb0d881) | ![WhatsApp Image 2023-08-02 at 22 34 39](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/08a44eff-67a7-4634-b7c1-50c718448b8d)

</details>

<details>
<summary>Extra Image</summary>

![WhatsApp Image 2023-08-02 at 22 24 48](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/ce63c82f-698c-435d-b3c8-7a544f5b26d1)

</details>

As I'm still a university student, I don't have much experience with Koltin, so I'd like to give most of the credit of this PR to @semenvav as I took his code from #9737 as reference to create this feature. 

If you notice any mistakes in the code or if you have a more efficient way of writing it, then please notify me so I can fix it.

AntsyLich commented on presentation-core/src/main/java/tachiyomi/presentation/core/screens/EmptyScreen.kt:
```diff
@@ -49,8 +51,9 @@ fun EmptyScreen(
     message: String,
     modifier: Modifier = Modifier,
     actions: List<EmptyScreenAction>? = null,
+    selector: Int = 1,
```
```suggestion
    happyFace: Boolean = false,
```

AntsyLich commented on presentation-core/src/main/java/tachiyomi/presentation/core/screens/EmptyScreen.kt:
```diff
@@ -49,8 +51,9 @@ fun EmptyScreen(
+    selector: Int = 1,
 ) {
-    val face = remember { getRandomErrorFace() }
+    val face = remember { getRandomFace(selector) }
```
```suggestion
    val face = remember { getRandomFace(happyFace) }
```

AntsyLich commented on presentation-core/src/main/java/tachiyomi/presentation/core/screens/EmptyScreen.kt:
```diff
@@ -36,8 +36,10 @@ fun EmptyScreen(
     @StringRes textResource: Int,
     modifier: Modifier = Modifier,
     actions: List<EmptyScreenAction>? = null,
+    selector: Int = 1,
```
```suggestion
    happyFace: Boolean = false,
```

AntsyLich commented on presentation-core/src/main/java/tachiyomi/presentation/core/screens/EmptyScreen.kt:
```diff
@@ -36,8 +36,10 @@ fun EmptyScreen(
+    selector: Int = 1,
 ) {
     EmptyScreen(
+        selector = selector,
         message = stringResource(textResource),
         modifier = modifier,
         actions = actions,
```
```suggestion
        message = stringResource(textResource),
        modifier = modifier,
        actions = actions,
        happyFace = happyFace,
```

AntsyLich commented on i18n/src/main/res/values/strings.xml:
```diff
@@ -898,6 +910,7 @@
 
     <!-- Information Text -->
     <string name="information_no_downloads">No downloads</string>
+    <string name="information_no_errors">No errors found</string>
```
```suggestion
    <string name="information_no_update_errors">All entries updated successfully</string>
```

AntsyLich commented on app/src/main/java/eu/kanade/tachiyomi/data/library/LibraryUpdateJob.kt:
```diff
@@ -550,6 +553,16 @@ class LibraryUpdateJob(private val context: Context, workerParams: WorkerParamet
          */
         private const val KEY_TARGET = "target"
 
+        private var newFailedUpdates: List<Pair<Manga, String?>> = emptyList()
+
+        fun getNewFailedUpdates(): List<Pair<Manga, String?>> {
+            return newFailedUpdates
+        }
+
+        fun setNewFailedUpdates(set: HashSet<Long>) {
+            newFailedUpdates = newFailedUpdates.filterNot { it.first.id in set }
+        }
```
So the errors will disappear after a cold start of app? Should store this info inside the db

Riztard commented:
![image](https://github.com/tachiyomiorg/tachiyomi/assets/16263232/d4e23582-4066-458d-90b0-7b74866ee928)
im wondering if we need to increase the error line to 2 for more visible details but looks bad 
or need to group it by error like the original to see more of the error detail and less repetitive/same error msg

Programmer-0-0 commented:
@Riztard I'm currently working to change the grouping to Source > errorMessage > Manga and yes the errorMessage will have multiple lines

AntsyLich commented on data/src/main/sqldelight/tachiyomi/data/failed_updates.sq:
```diff
@@ -0,0 +1,24 @@
+DELETE FROM failed_updates;
+
+getFailedUpdatesCount:
+SELECT COUNT(*) AS row_count FROM failed_updates;
```
new line

Programmer-0-0 replied:
you mean add a new line? not sure what you exactly mean tbh

AntsyLich replied:
Yes

AntsyLich commented on data/src/main/sqldelight/tachiyomi/data/failed_updates.sq:
```diff
@@ -0,0 +1,24 @@
+CREATE TABLE failed_updates (
+    manga_id INTEGER NOT NULL,
+    error_message TEXT,
```
```suggestion
    error_message TEXT,
    UNIQUE (manga_id) ON CONFLICT REPLACE,
```

AntsyLich commented on data/src/main/sqldelight/tachiyomi/migrations/26.sqm:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE IF NOT EXISTS failed_updates (
+    manga_id INTEGER NOT NULL,
+    error_message TEXT,
```
```suggestion
    error_message TEXT,
    UNIQUE (manga_id) ON CONFLICT REPLACE,
```

AntsyLich commented on data/src/main/sqldelight/tachiyomi/migrations/26.sqm:
```diff
@@ -0,0 +1,8 @@
+    ON DELETE CASCADE
+);
+
+CREATE INDEX IF NOT EXISTS idx_failed_updates ON failed_updates (manga_id);
```
Not needed cause of the unique clause above. Also if we needed index you would've needed to also create it inside `failed_updates.sq`
```suggestion
```

AntsyLich commented on data/src/main/java/tachiyomi/data/failed/FailedUpdatesRepositoryImpl.kt:
```diff
@@ -0,0 +1,45 @@
+        return handler.subscribeToList { failed_updatesQueries.getFailedUpdates(failedUpdatesMapper) }
+    }
+
+    override fun getFailedUpdatesCount(): Flow<Long> {
+        return handler.subscribeToOne { failed_updatesQueries.getFailedUpdatesCount() }
+    }
```
```suggestion
    override fun getFailedUpdatesCount(): Flow<Boolean> {
        return handler
            .subscribeToOne { failed_updatesQueries.getFailedUpdatesCount() }
            .map { it > 0 }
            .distinctUntilChanged()
    }
```

Programmer-0-0 replied:
This is actually quite a smart way to do it.

AntsyLich commented on data/src/main/java/tachiyomi/data/failed/FailedUpdatesRepositoryImpl.kt:
```diff
@@ -0,0 +1,45 @@
+class FailedUpdatesRepositoryImpl(
+    private val handler: DatabaseHandler,
+) : FailedUpdatesRepository {
+    override suspend fun getFailedUpdates(): Flow<List<FailedUpdate>> {
```
You can use a normal function?

Programmer-0-0 replied:
do you mean without suspend?

AntsyLich replied:
Yes

Programmer-0-0 replied:
I mean couldn't it work both ways?

AntsyLich replied:
You need a coroutine scope to execute suspend function. So if possible don't use suspend.

AntsyLich commented on data/src/main/java/tachiyomi/data/failed/FailedUpdatesRepositoryImpl.kt:
```diff
@@ -0,0 +1,45 @@
+        } catch (e: Exception) {
+            logcat(LogPriority.ERROR, e)
+        }
+    }
```
```suggestion
    }
    
```

Programmer-0-0 replied:
are you removing the catch statement over here. based on the other repositories I saw them using it so I thought of it more of a safe measure no?

AntsyLich replied:
Adding a line after

AntsyLich commented on data/src/main/java/tachiyomi/data/failed/FailedUpdatesRepositoryImpl.kt:
```diff
@@ -0,0 +1,45 @@
+            logcat(LogPriority.ERROR, e)
+        }
+    }
+    override suspend fun insert(manga: Manga, errorMessage: String?) {
+        handler.await(inTransaction = true) {
+            failed_updatesQueries.insert(
+                mangaId = manga.id,
+                errorMessage = errorMessage,
+            )
+        }
+    }
```
And why is error message nullable? Update the sql schema too to make it not nullable
```suggestion
    override suspend fun insert(mangaId: Long, errorMessage: String?) {
        handler.await(inTransaction = true) {
            failed_updatesQueries.insert(
                mangaId = mangaId,
                errorMessage = errorMessage,
            )
        }
    }
```

AntsyLich commented on data/src/main/java/tachiyomi/data/failed/FailedUpdatesMapper.kt:
```diff
@@ -0,0 +1,11 @@
+
+import tachiyomi.domain.failed.model.FailedUpdate
+
+val failedUpdatesMapper: (Long, String?) -> FailedUpdate = {
+        mangaId, errorMessage ->
```
```suggestion
val failedUpdatesMapper: (Long, String?) -> FailedUpdate = { mangaId, errorMessage ->
```

AntsyLich commented on app/src/main/java/eu/kanade/tachiyomi/ui/updates/UpdatesScreenModel.kt:
```diff
@@ -370,6 +378,7 @@ class UpdatesScreenModel(
         val isLoading: Boolean = true,
         val items: List<UpdatesItem> = emptyList(),
         val dialog: Dialog? = null,
+        val warningIconState: Boolean = false,
```
```suggestion
        val hasFailedUpdates: Boolean = false,
```

T3sT3ro commented on app/src/main/java/eu/kanade/presentation/updates/failed/FailedUpdatesScreen.kt:
```diff
@@ -57,23 +56,19 @@ import tachiyomi.presentation.core.screens.LoadingScreen
         val state by screenModel.state.collectAsState()
 
-        val categoryExpandedMapSaver: Saver<MutableMap<String, Boolean>, *> = Saver(
+        val categoryExpandedMapSaver: Saver<MutableMap<Pair<String, String?>, Boolean>, *> = Saver(
```
`Pair<String, String?>` is a bit cryptic, wouldn't domain specific `data class` be better?

something like `data class SomeError(val category: string, val errorMessage: String?)`

or whatever it is.

Programmer-0-0 replied:
@T3sT3ro tbh I'm not quite sure what you exactly mean but regardless I will remove it anyways since based on the new suggestions from arkon. I'll be removing the grouping by source and category options as well as expand and collapse ability. It will be grouped by error message instead.

T3sT3ro replied:
See discord, [tldr is here](https://softwareengineering.stackexchange.com/questions/359948/primitive-vs-class-to-represent-simple-domain-object)

Programmer-0-0 commented:
I did some new changes:
- Removed grouping by category
- Removed migrate button
- Improved UI look
- Added ability to copy error message (either by pressing on warning icon or long click)

Here is a video showing me playing around with it:

<details>
<summary>Show video</summary>

https://github.com/tachiyomiorg/tachiyomi/assets/67505416/f92d8074-6604-4f20-b23b-84bf1b49a099



</details>

T3sT3ro commented:
I like it. I would also think about: 
- adding red tint to the count badge (2/24) of failed updates
- maybe also adding an icon inside that pill that symbolizes error (also red, warning icon? exclamation mark in the triangle/circle?). See how it looks, it may look cluttered or may not, but it would be best to see for yourself
- the bottom padding of the source card seems to be messed up (manga icons are too close to the bottom border)
- show how a source card looks expanded when there are multiple error types. If it looks OK, then good, otherwise maybe think about adding some spacer above / bottom border right under the error message (to make it look more like an accordion) and a slight indent for manga titles (indent might looks bad and waste space, idk)?
- Maybe add a new preference (enabled by default) to always expand manga list for error in the source card group

arkon commented:
Since you mentioned you had some followup work on translating some exceptions to more user-friendly messages, just an FYI that we do have a mechanism for that here as seen here: 6922792ad110ce9194aa963674bbf2c5c1351ea1

Could potentially do a smaller separate PR for that first if you want to expand on that stuff.

Programmer-0-0 commented:
> Since you mentioned you had some followup work on translating some exceptions to more user-friendly messages, just an FYI that we do have a mechanism for that here as seen here: [6922792]
Ok so now I'm unsure how to go about this. The best solution I can think of is to remove the isOnline from database and revert to the older commit where I save the simplifiedErrorMessage instead that way I can pass it the screen model. please correct me if this isn't the correct way to go about it, else I'll proceed with making a separate smaller PR (for error message strings) and then adjust the changes here accordingly
